Add support for SHA512 in HTLCs

jsdanielh opened the pull request:
Add support for the SHA512 algorithm in HTLCs.

## Pull request checklist

- [X] All tests pass. The project builds and runs.
- [X] I have resolved any merge conflicts.
- [X] I have resolved all `clippy` and `rustfmt` warnings.

jsdanielh committed:
Add support for SHA512 in HTLCs

- Add support for the SHA512 algorithm in HTLCs.
- Remove the `HashAlhorithm` `enum` and moved its variants into
  `AnyHash`.

sisou reviewed (commented):
I see you are adding an additional byte to the HTLC that just encodes `00` for a 32-byte hash and `01` for a 64-byte hash. Is that really necessary? The length of the hash can be determined from the hash algorithm field in the previous byte, like in Nimiq 1.0. Or is this necessary for Rust serialization?

jsdanielh commented:
> I see you are adding an additional byte to the HTLC that just encodes `00` for a 32-byte hash and `01` for a 64-byte hash. Is that really necessary? The length of the hash can be determined from the hash algorithm field in the previous byte, like in Nimiq 1.0. Or is this necessary for Rust serialization?

Yeah, unfortunately it's necessary for serialization. But there is another solution which would imply removing the `hashAlgorithm` `enum` and encode it into `AnyHash`. This would be more efficient in terms of the space indeed. Let me come up with a possible implementation of this.

jsdanielh commented:
@sisou Ok, done. With this implementation the `data` field length will remain invariant in terms of the size for 32 byte hashes. The proof will increase in a byte, but it's better than the previous change where it was increasing in 2.

sisou commented on primitives/transaction/tests/htlc_contract_verify.rs:
```diff
@@ -128,70 +127,63 @@ fn it_can_verify_regular_transfer() {
-        )),
-        pre_image: AnyHash::from([0u8; 32]),
+        hash_root: AnyHash::from(Blake2bHasher::default().digest(&[0u8; 32])),
+        pre_image: AnyHash::Blake2b(AnyHash32::from([0u8; 32])),
```
Technically, the preimage is not necessarily a hash of any kind. It's just random data, not even technically a _hash_ at all. A secret basically.

jsdanielh replied:
Yeah, this is not elegant. One solution could be to use another type for the `pre_image` instead of `AnyHash`.

sisou commented on primitives/transaction/tests/htlc_contract_verify.rs:
```diff
@@ -128,70 +127,63 @@ fn it_can_verify_regular_transfer() {
-        hash_root: AnyHash::from(<[u8; 32]>::from(Sha256Hasher::default().digest(&[0u8; 32]))),
-        pre_image: AnyHash::from([0u8; 32]),
+        hash_root: AnyHash::from(Sha256Hasher::default().digest(&[0u8; 32])),
+        pre_image: AnyHash::Sha256(AnyHash32::from([0u8; 32])),
```
Should also add a test for sha512 I think.

sisou commented on rpc-interface/src/consensus.rs:
```diff
@@ -114,7 +114,6 @@ pub trait ConsensusInterface {
         htlc_recipient: Address,
         hash_root: AnyHash,
         hash_count: u8,
-        hash_algorithm: HashAlgorithm,
```
How are these RPC methods going to work? Does the `hash_root` field (hex string over JSON) now have to be prefixed by a byte representing the algo, so needs to be 33 bytes long?

jsdanielh replied:
No, this should work the same way we send `enums`. A tag field must be inserted to specify the `hashAlgorithm`. This is why I added [this](https://github.com/nimiq/core-rs-albatross/pull/1668/files#diff-d1b20c99015ddf82b249f66f2236dd9541bfdbba5b728f8eae881b34bdff1436R73).

sisou replied:
OK, I tested this. And while I think that it's not very ergonomic, it does work.

Still, for the `preimage` setting the hash algo doesn't make much sense usually.

jsdanielh replied:
@sisou Moved the `pre_image` to a separate type

sisou commented on rpc-interface/src/types.rs:
```diff
@@ -95,16 +93,7 @@ impl FromStr for ValidityStartHeight {
-            HashAlgorithm::Sha256 => HTLCContractHashAlgorithm::Sha256,
-        }
-    }
+    Sha512 = 4,
```
`HashAlgorithm` is unused now?

jsdanielh replied:
Its actually used by the `rpc-client` [here](https://github.com/nimiq/core-rs-albatross/pull/1668/files#diff-c37ba4db67fea5930db56228ef619010a93f6144fb2d4427c61e8992d5a4346bR10).

jsdanielh committed:
htlc: Add the `PreImage` type

- Add the `PreImage` type for HTLC pre images instead of using
  `AnyHash` that now requires the algorithm to be set.
- Manually implement the serialization and deserialization for
  `AnyHash` and `PreImage` such that they can be implemented
  differently if the serialization is human readable (json) or not.

sisou commented on primitives/transaction/src/account/htlc_contract.rs:
```diff
@@ -269,52 +323,219 @@ impl OutgoingHTLCTransactionProof {
+                let mut state = serializer.serialize_struct("PreImage", PREIMAGE_FIELDS.len())?;
+                match self {
+                    PreImage::PreImage32(pre_image) => {
+                        state.serialize_field(PREIMAGE_FIELDS[0], &0u8)?;
```
I believe it's usual for the byte to express the length of the following field. So this number could be `32u8`, and for `PreImage64` it could be `64u8`. I think that makes more sense than `0` and `1`.

jsdanielh replied:
Ok, yeah, for vectors it is true. I will change it

jsdanielh replied:
Done @sisou

sisou commented on rpc-interface/src/consensus.rs:
```diff
@@ -138,7 +138,7 @@ pub trait ConsensusInterface {
         contract_address: Address,
         recipient: Address,
-        pre_image: AnyHash,
+        pre_image: PreImage,
```
So do I understand the deserializer for `PreImage` correctly, that in the JSON-RPC call just a regular HEX string is now expected for the `pre_image` field?

jsdanielh replied:
Exactly!
